design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,272 @@
+* Version: The API version of the group (e.g., `v1alpha1`)
+* Kind: The specific type within the API group (e.g., `cluster`)
+
+Putting this together with an example, the `cluster` kind from the `rook.io` API group with a version of `v1alpha1` would be referred to in full as `cluster.rook.io/v1alpha1`.
+
+#### StorageScope
+
+The `storageScope.rook.io/v1alpha2` type defines the boundaries or "scope" of the resources that comprise the backing storage substrate for a cluster.
+This could be devices, filters, directories, nodes, [persistent volumes](https://github.com/rook/rook/issues/919), and others.
+There are user requested means of selecting storage that Rook doesn't currently support that could be included in this type, such as the ability to select a device by path instead of by name, e.g. `/dev/disk/by-id/`.
+Also, wildcards/patterns/globbing should be supported on multiple resource types, removing the need for the current `useAllNodes` and `useAllDevices` boolean fields.
+
+By encapsulating this concept as its own type, it can be reused within other custom resources of Rook.
+For instance, this would enable Rook to support storage of other types that could benefit from orchestration in cloud-native environments beyond distributed storage systems.
+`StorageScope` could also provide an abstraction from such details as device name changes across reboots.
+
+#### StorageCluster
+
+The `storageCluster.rook.io/v1alpha2` type provides the general abstraction for a distributed storage cluster.
+The spec for this type will include image version, networking information, placement information, resource usage requests, etc.
+Backend specific fields, such as Ceph configuration, will be included under a field identified by the backend's name.
+Examples of the `StorageCluster` type are included in the [examples](#examples) section.
+
+In addition to the existing support for Ceph, we should consider support for [Minio](https://www.minio.io/) as the next storage backend supported by Rook.
+Minio is a popular distributed object storage solution that could benefit from a custom controller to orchestrate, configure and manage it in cloud-native environments.
+
+#### Additional Types
+
+Rook and the greater community would also benefit from additional types and abstractions.
+We should work on defining those further, but it is out of scope for this design document that is focusing on support for multiple storage backends.
+Some potential ideas for additional types to support in Rook:
+
+* [Snapshots, backup and policy](https://github.com/rook/rook/issues/1552)
+* Quality of Service (QoS), resource consumption (I/O and storage limits)
+
+## Examples
+
+This section contains concrete examples of storage clusters as a user would define them using yaml.
+In addition to the Ceph and Minio examples, we will be considering support for additional storage types in the near future.
+
+`ceph-cluster.yaml`:
+
+```yaml
+apiVersion: rook.io/v1alpha2
```

Comment:
Backend versioning is another sticky issue.  In the situation you describe, where we have a single StorageCluster CRD for all backends, I would imagine that the new backend `foo` would be added to the StorageCluster type in the alpha group, not to the one in the stable group.  I don't **think** that breaks any versioning principles.

But would that mean that the new `foo` backend only gets access to the older alpha version of `StorageScope`? I don't think it would get to use the newest/stable `StorageScope` spec, unless we bumped alpha to `v1alpha2` and that `v1alpha1` version of the `StorageCluster` could reference the latest `v1stable1` version of `StorageScope`.  That would not break folks that are using the `v1alpha1` `StorageCluster`.

I believe this same exact problem occurs whether we have backend specific CRDs or a single cluster CRD with subsections for specific backends.  To maybe make this discussion more visual, we're deciding between the following:
**Common Cluster CRD**
```
apiVersion: rook.io/v1alpha1
kind: StorageCluster
spec:
  ceph:
    monCount: 3
---
apiVersion: rook.io/v1alpha1
kind: StorageCluster
spec:
  minio:
    mode: distributed
```

**Backend specific cluster CRDs**
```
apiVersion: rook.io/v1alpha1
kind: Ceph
spec:
  monCount: 3
---
apiVersion: rook.io/v1alpha1
kind: Minio
spec:
  mode: distributed
```

It **feels** more natural for when a new (alpha) backend is added, that it would be its own type that is added to the `alpha` group, as opposed to updating the `alpha` group's `StorageCluster` type to now support a new subsection for the new backend.

But I'm also concerned about reuse of common cluster properties if each backend is it's own cluster CRD type.  The story for common cluster properties gets weaker there.